include/lsst/daf/base/DateTime.h

```diff
@@ -57,44 +61,230 @@ namespace base {
+
+    /**
+     * Default constructor: construct an invalid DateTime
+     */
+    explicit DateTime();
+
+    /**
+     * Construct a DateTime from nanoseconds since the unix epoch
+     *
+     * @param[in] nsecs  integer nanoseconds since the unix epoch;
+                         if nsecs == DateTime.invalid_nsecs then the DateTime is invalid,
+                         regardless of scale
+     * @param[in] scale  time scale of input (TAI, TT or UTC, default TAI).
+     * @throw lsst.pex.exceptions.DomainError if scale is UTC and the date is before 1961-01-01
+     */
+    explicit DateTime(long long nsecs, Timescale scale=TAI);
+
+    /**
+     * Construct a DateTime from a double in the specified system and scale
+     *
+     * @param[in] date  specified date
+     * @param[in] system  time system of input (JD, MJD, or EPOCH)
+     * @param[in] scale  time scale of input (TAI, TT or UTC, default TAI).
+     * @throw lsst.pex.exceptions.DomainError if scale is UTC and the date is before 1961-01-01
+     */
+    explicit DateTime(double date, DateSystem system=MJD, Timescale scale=TAI);
+
+    /**
+     * Construct a DateTime from year, month, day, etc. (tm struct fields)
+     *
+     * @param[in] year   year; must be in the range [1902, 2261], inclusive.
+     * @param[in] month  month number, where January = 1
+     * @param[in] day  day of the month (1 to 31).
+     * @param[in] hr  hour (0 to 23)
+     * @param[in] min  minute (0 to 59)
+     * @param[in] sec  integer seconds (0 to 60)
+     * @param[in] scale  time scale of input (TAI, TT or UTC, default TAI).
+     * @throw lsst.pex.exceptions.DomainError if scale is UTC and the date is before 1961-01-01,
+     *      or the year < 1902 or > 2261
+     */
+    DateTime(int year, int month, int day, int hr, int min, int sec, Timescale scale=TAI);
+
+    /**
+     * Construct a DateTime from an ISO8601 string
+     *
+     * Accepts a restricted subset of ISO8601:
+     *     yyyy-mm-ddThh:mm:ss.nnnnnnnnnZ
+     * where:
+     * - the final Z is required for UTC and prohibited for TAI or TT
+     * - the - and : separators are optional
+     * - the decimal point and fractional seconds are optional
+     * - the decimal point may be a comma
+     *
+     * @param[in] iso8601  ISO8601 string representation of date and time
+     * @param[in] scale  time scale of input (TAI, TT or UTC, default TAI).
+     * @throw lsst.pex.exceptions.DomainError if scale is UTC and the date is before 1961-01-01,
+     *      or the year < 1902 or > 2261 or iso8601 is not in an acceptable format
+     */
     explicit DateTime(std::string const& iso8601, Timescale scale);
 
-    long long nsecs(Timescale scale = TAI) const;
-    double get(DateSystem system = MJD, Timescale scale = TAI) const;
+    /**
+     * Get date as nanoseconds since the unix epoch
+     *
+     * @note If the DateTime is invalid then the returned value is DateTime.invalid_nsecs,
+     * regardless of scale.
```

Comment:
I think there is exactly one situation where round tripping can fail; either there is some date for which DateTime(special_date, scale=TT) will produce an invalid date because the internal nsecs will be invalid_nsecs **or else** a special  DateTime for which get(TT) will return invalid_nsecs. It does not occur for UTC because UTC dates cannot be that old. It also does not occur for ISO dates nor dates constructed from `tm struct` fields because those cannot be old enough. I don't think it is worth worrying about, since the particular date is ancient and obscure, but I suppose we could try to enforce some kind of range limit for the MJD and nsecs constructors.

Other than that, I think round tripping is robust because the nsecs constructor and `get` both ignore the scale argument if nsecs == invalid_nsecs.